### PR TITLE
chore: updating storybook-react to use the tailwind preset package and yarn config updates

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This PR updates the `storybook-react` setup in the `metamask-design-system` monorepo to integrate the `@metamask/design-system-tailwind-preset` package. This enables more effective testing of the Tailwind preset in a Storybook environment, reflecting its use in both the MetaMask Portfolio and Extension projects.

While working on this, I identified the need to adjust the `yarn.config` to accommodate non-publish private package constraints, which apply to both the `storybook-react` and `storybook-react-native` applications. Additionally, updates to `tsconfig` and ESLint settings were necessary to ensure correct resolution of workspace package imports. We should also adopt workspace versioning for dependencies, similar to the approach used in the `snaps` monorepo.

### **Key Enhancements**
  - Integrated `@metamask/design-system-tailwind-preset` into the `storybook-react` configuration.
  - Updated `yarn.config` to support non-publish private package constraints for `storybook-react` and `storybook-react-native`.
  - Adjusted `tsconfig.packages.json` to include required paths for workspace package imports.
  - Improved ESLint settings to resolve workspace package imports, addressing `import/no-unresolved` errors.
  - Adopted workspace versioning strategies similar to the `snaps` monorepo for consistent dependency management.
  - Refactored the example `Button` story to temporarily use Tailwind-generated classnames from the new preset.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-design-system/issues/63 https://github.com/MetaMask/metamask-design-system/issues/62

## **Manual testing steps**

1. Install the dependencies in the monorepo.
2. Start the `storybook-react` application.
3. Verify that the `Button` story reflects the new Tailwind classnames.
4. Test the integration of the Tailwind preset by adding additional design token classnames from the preset to components in Storybook.

### Before

https://github.com/user-attachments/assets/14ed072b-fe38-4f85-94a1-473bb83c834d

### After

https://github.com/user-attachments/assets/66f7ab18-15c7-411b-a067-bcff6ef13c7b

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs).
- [x] I've completed the PR template to the best of my ability.
- [x] I’ve included tests if applicable.
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable.
- [x] I’ve applied the correct labels to the PR.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR.
- [ ] I confirm that the PR addresses all acceptance criteria.
- [ ] Testing evidence (e.g., recordings, screenshots) is included where applicable.

### CODE REVIEW

File: apps/storybook-react/.storybook/theme.ts [hunk 1]
```diff
@@ -3,7 +3,6 @@ import { create } from '@storybook/theming';
 export default create({
   base: 'light',
   brandTitle: 'MetaMask Design System React UI',
-  brandUrl: 'https://metamask.io/',
```
Comment (georgewrmarshall) [inline apps/storybook-react/.storybook/theme.ts:6]: I don't think there is any benefit in the logo going to the marketing site.

File: apps/storybook-react/package.json [hunk 2]
```diff
@@ -1,6 +1,5 @@
 {
   "name": "@metamask/storybook-react",
-  "version": "0.0.0",
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:3]: No need for a version as this package is not going to be published.

File: apps/storybook-react/package.json [hunk 3]
```diff
@@ -9,6 +8,7 @@
   },
   "devDependencies": {
     "@chromatic-com/storybook": "^1.9.0",
+    "@metamask/design-system-tailwind-preset": "workspace:^",
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:11]: Adding the tailwind preset package

File: apps/storybook-react/package.json [hunk 4]
```diff
@@ -31,7 +31,6 @@
     "typescript": "~5.2.2",
     "vite": "^5.4.8"
   },
-  "packageManager": "yarn@4.2.2",
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:34]: This is managed at the root `package.json`

File: apps/storybook-react/stories/Button.stories.ts [hunk 5]
```diff
@@ -5,49 +5,18 @@ import { Button } from './Button';
...
-};
-
-export const Small: Story = {
+export const Default: Story = {
   args: {
-    size: 'small',
-    label: 'Button',
+    children: 'Button',
   },
 };
```
Comment (georgewrmarshall) [inline apps/storybook-react/stories/Button.stories.ts:22]: Simplifying storybook story to just one button story

File: apps/storybook-react/stories/Button.tsx [hunk 6]
```diff
@@ -1,41 +1,18 @@
...
-  );
-};
+export const Button = ({ children, ...props }: ButtonProps) => (
+  <button
+    className="px-4 h-8 rounded-full bg-primary-default text-primary-inverse"
+    {...props}
+  >
+    {children}
+  </button>
+);
```
Comment (georgewrmarshall) [inline apps/storybook-react/stories/Button.tsx:18]: Simplifying component to use Tailwind classnames

File: apps/storybook-react/stories/button.css
Comment (georgewrmarshall) [inline apps/storybook-react/stories/button.css:1]: Removing unused CSS file

File: apps/storybook-react/tsconfig.json [hunk 7]
```diff
@@ -1,7 +1,8 @@
 {
   "extends": "../../tsconfig.packages.json",
   "compilerOptions": {
-    "baseUrl": "./"
+    "baseUrl": "./",
+    "jsx": "react"
```
Comment (georgewrmarshall) [inline apps/storybook-react/tsconfig.json:5]: ### Before

Screenshot 2024-10-23 at 2 42 56 PM

### After

Screenshot 2024-10-23 at 2 43 02 PM

File: apps/storybook-react/tailwind.config.js [hunk 8]
```diff
@@ -1,12 +1,12 @@
...
+const designSystemPreset = require('@metamask/design-system-tailwind-preset');
+
 /** @type {import('tailwindcss').Config} */
 module.exports = {
+  presets: [designSystemPreset],
   content: [
     // '../../packages/react-components/src/**/*.{js,jsx,ts,tsx}', TO DO: Update when react component package is created
     './src/**/*.{js,jsx,ts,tsx}',
-    './.storybook/**/*.{js,jsx,ts,tsx}',
+    './stories/**/*.{js,jsx,ts,tsx}',
```
Comment (georgewrmarshall) [inline apps/storybook-react/tailwind.config.js:10]: Making sure that we are getting tailwind to read the correct files

File: apps/storybook-react-native/package.json [hunk 9]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "@metamask/storybook-react-native",
-  "version": "0.0.0",
   "private": true,
   "description": "Storybook React Native",
+  "files": [],
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/package.json:5]: Part of the yarn constraints that we do want private non publishable packages to adhere to https://github.com/MetaMask/metamask-design-system/blob/main/yarn.config.cjs#L172-L176

File: apps/storybook-react-native/package.json [hunk 10]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "@metamask/storybook-react-native",
-  "version": "0.0.0",
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/package.json:3]: version number isn't required as this package will not be published

File: apps/storybook-react/package.json [hunk 11]
```diff
@@ -1,14 +1,15 @@
 {
   "name": "@metamask/storybook-react",
-  "version": "0.0.0",
   "private": true,
   "description": "storybook react",
+  "files": [],
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:5]: Part of the yarn constraints that we do want private non publishable packages to adhere to. See this comment in `yarn.config.cjs` https://github.com/MetaMask/metamask-design-system/blob/main/yarn.config.cjs#L172-L176

File: apps/storybook-react/tailwind.config.js [hunk 12]
```diff
@@ -1,12 +1,11 @@
+const designSystemPreset = require('@metamask/design-system-tailwind-preset');
```
Comment (georgewrmarshall) [inline apps/storybook-react/tailwind.config.js:1]: Importing tailwind preset for use in tailwind config

File: yarn.config.cjs [hunk 13]
```diff
@@ -43,15 +43,10 @@ module.exports = defineConfig({
         Yarn.dependencies({ workspace }),
       );
 
-      // Completely ignore apps workspaces
-      if (isApps) {
-        continue;
-      }
-
```
Comment (georgewrmarshall) [inline yarn.config.cjs:50]: This ignored ever check which may have been a bit too broad. There are checks for `private` packages that we should adhere to

File: apps/storybook-react/package.json [hunk 14]
```diff
@@ -1,14 +1,15 @@
...
   "scripts": {
     "build-storybook": "storybook build",
     "storybook": "storybook dev -p 6006"
   },
   "devDependencies": {
     "@chromatic-com/storybook": "^1.9.0",
+    "@metamask/design-system-tailwind-preset": "^0.0.0",
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:12]: We have the option of updating to use "workspace:^" here when we are developing but there is a yarn constraint to make sure every package uses the dependencies latest version which I think is a good check. See comment in `yarn.config.cjs` https://github.com/MetaMask/metamask-design-system/blob/main/yarn.config.cjs#L577-L600

File: tsconfig.json [hunk 15]
```diff
@@ -7,6 +7,5 @@
     { "path": "./apps/storybook-react-native" },
     { "path": "./apps/storybook-react" }
   ],
-  "files": [],
-  "include": ["./types"]
```
Comment (georgewrmarshall) [inline tsconfig.json:11]: the `"./types"` directory is left over from core which the metamask-design-system doesn't have

Screenshot 2024-10-23 at 8 06 22 PM

File: apps/storybook-react/package.json [hunk 16]
```diff
@@ -1,14 +1,15 @@
...
   "private": true,
   "description": "storybook react",
+  "files": [],
   "scripts": {
     "build-storybook": "storybook build",
     "storybook": "storybook dev -p 6006"
   },
   "devDependencies": {
     "@chromatic-com/storybook": "^1.9.0",
+    "@metamask/design-system-tailwind-preset": "workspace:*",
```
Comment (georgewrmarshall) [inline apps/storybook-react/package.json:12]: Adding `@metamask/design-system-tailwind-preset` as a dev dependency that points to the workspace

File: yarn.config.cjs [hunk 17]
```diff
@@ -577,8 +572,7 @@ function expectCorrectWorkspaceChangelogScripts(workspace) {
 /**
  * Expect that if the workspace package lists another workspace package within
  * `dependencies` or `devDependencies`, the version used within the dependency
- * range is exactly equal to the current version of the dependency (and the
- * range uses the `^` modifier).
+ * is `workspace:^`.
```
Comment (georgewrmarshall) [inline yarn.config.cjs:575]: An update that enforces all workspace package dependencies to use the workspace version. This is automatically updated using `yarn publish` and aligns with how Snaps approaches workspace package dependencies: https://github.com/MetaMask/snaps/blob/main/yarn.config.cjs#L650-L672

File: apps/storybook-react/tailwind.config.js [hunk 18]
```diff
@@ -1,12 +1,12 @@
+// eslint-disable-next-line import/no-unresolved
```
Comment (georgewrmarshall) [inline apps/storybook-react/tailwind.config.js:1]: Need to figure out workspace package imports in the apps/ workspace currently we are running into this error 

```Error:   1:36  error  Unable to resolve path to module '@metamask/design-system-tailwind-preset'  import/no-unresolved```

File: .eslintrc.js [hunk 19]
```diff
@@ -67,7 +67,28 @@ module.exports = {
...
+        // Specify file extensions that ESLint should consider when resolving modules
+        extensions: ['.js', '.jsx', '.ts', '.tsx'],
+        // Include workspace directories to help ESLint resolve modules within these folders
+        paths: ['apps', 'packages'],
+        /**
+         * The 'node' resolver is configured to recognize JavaScript and TypeScript file extensions.
+         * The 'paths' array includes 'apps' and 'packages' directories, aligning with the workspace structure.
+         * This ensures that modules imported from these directories are correctly resolved by ESLint.
+         */
+      },
```
Comment (georgewrmarshall) [inline .eslintrc.js:91]: Adding in depth comments here

File: .eslintrc.js [hunk 20]
```diff
@@ -67,7 +67,28 @@ module.exports = {
...
+         * The 'paths' option allows ESLint to understand module aliases defined in TypeScript.
+         * Here, 'apps/*' and 'packages/*' correspond to the directories where your workspace packages reside.
+         * This should align with the 'paths' defined in your TypeScript compilerOptions.
+         */
+      },
+      node: {
+        // Specify file extensions that ESLint should consider when resolving modules
+        extensions: ['.js', '.jsx', '.ts', '.tsx'],
+        // Include workspace directories to help ESLint resolve modules within these folders
+        paths: ['apps', 'packages'],
```
Comment (georgewrmarshall) [inline .eslintrc.js:85]: The 'node' resolver is configured to recognize JavaScript and TypeScript file extensions. The 'paths' array includes 'apps' and 'packages' directories, aligning with the workspace structure. This ensures that modules imported from these directories are correctly resolved by ESLint.

File: .eslintrc.js [hunk 21]
```diff
@@ -67,7 +67,28 @@ module.exports = {
   },
   settings: {
     'import/resolver': {
-      typescript: {},
+      typescript: {
+        // Specify the path to the TypeScript configuration file
+        project: './tsconfig.packages.json',
+        // Define additional paths for module resolution to match TypeScript paths
+        paths: ['apps/*', 'packages/*'],
```
Comment (georgewrmarshall) [inline .eslintrc.js:74]: The 'paths' option allows ESLint to understand module aliases defined in TypeScript. Here, 'apps/*' and 'packages/*' correspond to the directories where your workspace packages reside.

File: tsconfig.packages.json [hunk 22]
```diff
@@ -13,7 +14,7 @@
      * `jest.config.packages.js`.
      */
     "paths": {
```
Comment (brianacnguyen) [inline tsconfig.packages.json:16]: i think you need to remove the comments here

**Review by brianacnguyen — APPROVED** [review]
non-blocking: remove comments from json file

File: tsconfig.packages.json (hunk 22 above)
Comment (georgewrmarshall) [reply tsconfig.packages.json:16]: Yes I noticed this! I realized there are a bunch of comments in the tsconfig files from core. So, maybe I didn't need to update the `tsconfig.packages.build.json` to `jsonc`. We can probably revert that to stay aligned with core.

**Pull request merged**